src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,520 @@
+			char text[35];
+			SAFE_SPRINTF(text, "Contains %d nested scripts", ctx->complexScripts[ctx->level].remainingScripts);
+
+			ui_displayPaginatedText(
+			        HEADER,
+			        text,
+			        this_fn
+			);
+		}
+		break;
+		case UI_SCRIPT_N_OF_K: {
+			{
+				// max possible length 67: "Reuqired signatures: x. Contains x nested scripts"
+				// where x is 2^32-1
+				char text[68];
+				SAFE_SPRINTF(text, "Required signatures: %u. Contains %u nested scripts", ctx->scriptContent.requiredScripts, ctx->complexScripts[ctx->level].remainingScripts);
+
+				ui_displayPaginatedText(
+				        HEADER,
+				        text,
+				        this_fn
+				);
+			}
+			break;
+		}
+		case UI_SCRIPT_INVALID_BEFORE:
+		case UI_SCRIPT_INVALID_HEREAFTER:
+			ui_displayUint64Screen(
+			        HEADER,
+			        ctx->scriptContent.timelock,
+			        this_fn
+			);
+			break;
+		default:
+			THROW(ERR_INVALID_STATE);
+		}
+	}
+
+	UI_STEP(DISPLAY_UI_STEP_RESPOND) {
+		io_send_buf(SUCCESS, NULL, 0);
+		ui_displayBusy();
+	}
+
+	UI_STEP_END(DISPLAY_UI_STEP_INVALID);
+}
+#undef HEADER
+#undef ASSERT_UI_SCRIPT_TYPE_SANITY
+
+#define UI_DISPLAY_SCRIPT(UI_TYPE)\
+	ctx->ui_scriptType = UI_TYPE;\
+	ctx->ui_step = DISPLAY_UI_STEP_POSITION;\
+	deriveScriptHash_display_ui_runStep();
+
+// Start complex native script
+
+static void deriveNativeScriptHash_handleAll(read_view_t* view)
+{
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE_WITH_CTX("");
+
+	nativeScriptHashBuilder_startComplexScript_all(&ctx->hashBuilder, ctx->complexScripts[ctx->level].remainingScripts);
+
+	ctx->complexScripts[ctx->level].ui_type = UI_SCRIPT_ALL;
+	UI_DISPLAY_SCRIPT(UI_SCRIPT_ALL);
+}
+
+static void deriveNativeScriptHash_handleAny(read_view_t* view)
+{
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE_WITH_CTX("");
+
+	nativeScriptHashBuilder_startComplexScript_any(&ctx->hashBuilder, ctx->complexScripts[ctx->level].remainingScripts);
+
+	ctx->complexScripts[ctx->level].ui_type = UI_SCRIPT_ANY;
+	UI_DISPLAY_SCRIPT(UI_SCRIPT_ANY);
+}
+
+static void deriveNativeScriptHash_handleNofK(read_view_t* view)
+{
+	// parse data
+	ctx->scriptContent.requiredScripts = parse_u4be(view);
+
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE_WITH_CTX("required scripts = %u, ", ctx->scriptContent.requiredScripts);
+
+	// validate that the received requiredScripts count makes sense
+	VALIDATE(ctx->complexScripts[ctx->level].remainingScripts >= ctx->scriptContent.requiredScripts, ERR_INVALID_DATA);
+
+	nativeScriptHashBuilder_startComplexScript_n_of_k(&ctx->hashBuilder, ctx->scriptContent.requiredScripts, ctx->complexScripts[ctx->level].remainingScripts);
+
+	ctx->complexScripts[ctx->level].ui_type = UI_SCRIPT_N_OF_K;
+	UI_DISPLAY_SCRIPT(UI_SCRIPT_N_OF_K);
+}
+
+static void deriveNativeScriptHash_handleComplexScriptStart(read_view_t* view)
+{
+	TRACE_BUFFER(view->begin, (size_t) (view->end - view->begin));
+
+	expectsMoreScripts();
```

Comment:
This is confusing. It will be much clearer if there was a bool function, say, `areMoreScriptsExpected`, and then
`VALIDATE(areMoreScriptsExpected() == true, ERR_INVALID_STATE)` here.